File: commons/ocaml/reader.ml
open Core

let get_year_day executable =
  let parts = String.split ~on:'/' executable in
  let name = List.last_exn parts in
  String.lsplit2_exn name ~on:'_'

let get_filename () =
  let test = ref false in
  let speclist = [ ("--test", Arg.Set test, "Test mode") ] in
  Arg.parse speclist print_endline "";
  if !test then "sample" else "data"

let get_filepath () =
  let year, day = get_year_day (Sys.get_argv ()).(0) in
  let filename = get_filename () in
  String.concat ~sep:"/" [ year; day; filename ^ ".txt" ]

let read () : string =
  let filepath = get_filepath () in
  In_channel.read_all filepath

let read_groups () : string list list =
  let data = read () in
  let groups = Str.split (Str.regexp "\n\n") data in
  List.map ~f:String.split_lines groups

let read_lines () : string list =
  let filepath = get_filepath () in
  In_channel.read_lines filepath

let read_ints () : int list =
  let lines = read_lines () in
  List.map ~f:int_of_string lines

let read_grid () : Grid.t =
  let lines = read_lines () in
  Grid.parse_grid lines

File: commons/rust/src/reader.rs
use crate::prelude::*;
use std::env;
use std::fmt::Debug;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::str::FromStr;

#[derive(Debug)]
struct Cli<'a> {
    file_name: &'a str,
}

impl Cli<'_> {
    fn parse() -> Self {
        let test = match env::args().nth(1) {
            None => false,
            Some(value) => value == "--test",
        };
        Self {
            file_name: if test { "sample.txt" } else { "data.txt" },
        }
    }

    fn file_path(&self) -> String {
        let (year, day) = Self::get_year_day();
        format!("data/{year}/{day}/{}", self.file_name)
    }

    fn get_year_day() -> (String, String) {
        let path = env::current_exe().unwrap();
        let binary = path.file_name().unwrap().to_str().unwrap();
        let parts: Vec<&str> = binary.split('_').collect();
        (parts[1].into(), parts[2].into())
    }
}

#[derive(Debug)]
pub struct Reader {
    path: String,
}

impl Default for Reader {
    fn default() -> Self {
        Self::new(Cli::parse().file_path())
    }
}

impl Reader {
    pub fn new(path: String) -> Self {
        Self { path }
    }

    pub fn grid<T: FromChar>(&self) -> Grid<T> {
        (&self.lines::<String>()).into()
    }

    pub fn chars<T: FromChar>(&self) -> Vec<T> {
        self.line::<String>()
            .chars()
            .map(|ch| T::from_char(ch).unwrap())
            .collect()
    }

    pub fn csv<T>(&self) -> Vec<T>
    where
        T: FromStr,
        T::Err: Debug,
    {
        self.line::<String>()
            .split(',')
            .map(|s| s.parse().unwrap())
            .collect()
    }

    pub fn groups<T>(&self) -> Vec<T>
    where
        T: FromStr,
        T::Err: Debug,
    {
        self.lines::<String>()
            .split(|line| line.is_empty())
            .map(|lines| lines.join("\n").parse().unwrap())
            .collect()
    }

    pub fn line<T>(&self) -> T
    where
        T: FromStr,
        T::Err: Debug,
    {
        self.lines().remove(0)
    }

    pub fn lines<T>(&self) -> Vec<T>
    where
        T: FromStr,
        T::Err: Debug,
    {
        match File::open(&self.path) {
            Ok(file) => BufReader::new(file)
                .lines()
                .map(|s| s.unwrap().parse().unwrap())
                .collect(),
            Err(_) => panic!("Could not open: {}", self.path),
        }
    }
}
